src/zkevm_specs/evm/util/call_gadget.py

```diff
@@ -0,0 +1,118 @@
+from zkevm_specs.evm.table import AccountFieldTag
+from zkevm_specs.util.arithmetic import RLC
+from zkevm_specs.util.hash import EMPTY_CODE_HASH
+from zkevm_specs.util.param import (
+    GAS_COST_ACCOUNT_COLD_ACCESS,
+    GAS_COST_CALL_WITH_VALUE,
+    GAS_COST_NEW_ACCOUNT,
+    GAS_COST_WARM_ACCESS,
+)
+from ...util import (
+    FQ,
+    N_BYTES_ACCOUNT_ADDRESS,
+    N_BYTES_GAS,
+    RLC,
+)
+from ..instruction import Instruction
+
+
+class CallGadget:
+    IS_SUCCESS_CALL: FQ
+
+    gas: FQ
+    callee_address: FQ
+    value: RLC
+    cd_offset: FQ
+    cd_length: FQ
+    rd_offset: FQ
+    rd_length: FQ
+    is_success: FQ
+
+    is_u64_gas: FQ
+    next_memory_size: FQ
+    memory_expansion_gas_cost: FQ
+
+    has_value: FQ
+    callee_code_hash: FQ
+    is_empty_code_hash: FQ
+
+    def __init__(
+        self,
+        instruction: Instruction,
+        is_success_call: FQ,
+        is_call: FQ,
+        is_callcode: FQ,
+        is_delegatecall: FQ,
+    ):
+        self.IS_SUCCESS_CALL = is_success_call
+
+        # Lookup values from stack
+        self.gas_rlc = instruction.stack_pop()
+        callee_address_rlc = instruction.stack_pop()
+        # For non-OOG case,
+        # the third stack pop `value` is not present for both DELEGATECALL and
+        # STATICCALL opcodes.
+        self.value = instruction.stack_pop() if is_call + is_callcode == FQ(1) else RLC(0)
+        cd_offset_rlc = instruction.stack_pop()
+        cd_length_rlc = instruction.stack_pop()
+        rd_offset_rlc = instruction.stack_pop()
+        rd_length_rlc = instruction.stack_pop()
+        self.is_success = instruction.stack_push().expr()
+
+        if self.IS_SUCCESS_CALL == FQ(1):
+            # Verify is_success is a bool
+            instruction.constrain_bool(self.is_success)
+            self.gas = instruction.rlc_to_fq(self.gas_rlc, N_BYTES_GAS)
+            self.is_u64_gas = instruction.is_zero(
+                instruction.sum(self.gas_rlc.le_bytes[N_BYTES_GAS:])
+            )
+        else:
+            instruction.constrain_zero(self.is_success)
+        self.has_value = FQ(0) if is_delegatecall == FQ(1) else 1 - instruction.is_zero(self.value)
+
+        self.callee_address = instruction.rlc_to_fq(callee_address_rlc, N_BYTES_ACCOUNT_ADDRESS)
+        self.cd_offset, self.cd_length = instruction.memory_offset_and_length(
+            cd_offset_rlc, cd_length_rlc
+        )
+        self.rd_offset, self.rd_length = instruction.memory_offset_and_length(
+            rd_offset_rlc, rd_length_rlc
+        )
+        # Verify memory expansion
+        (
+            self.next_memory_size,
+            self.memory_expansion_gas_cost,
+        ) = instruction.memory_expansion_dynamic_length(
+            self.cd_offset,
+            self.cd_length,
+            self.rd_offset,
+            self.rd_length,
+        )
```

Comment:
I noticed that the fields `callee_code_hash` and `is_empty_code_hash` are not set in the `__init__`.  They are set in the `gas_cost` method.  Why is that?  Isn't it better to initialize all the Class fields in `__init__`?  The circuit does it like this;
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/b77613e9b42eacb03ccc894f0b353c374da3f283/zkevm-circuits/src/evm_circuit/util/common_gadget.rs#L523-L525

Reply:
yes, I totally agree that. However, I don't know how to hadle it properly because I need to get `callee_code_hash` first as following,
```
        self.callee_code_hash = instruction.account_read(
            self.callee_address, AccountFieldTag.CodeHash
        )
```

If I moved above code to `__init__`, it breaks the order of rw table. (reading the value of `AccountFieldTag.CodeHash` must be after reading the value of `AccountFieldTag.Balance`, see [here](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_callop.py#L345-L355)). Either I change the order in the test or assign an arbitrary value in `__init__`(It will fail if anyone uses these vars before calling `gas_cost`, but I don't think it's a good idea). Do you have any better ideas for this?

Reply:
Ah I see!
After this change was applied https://github.com/privacy-scaling-explorations/zkevm-specs/pull/354 it will become impossible to read a `AccountFieldTag.Balance` for non-existing accounts.  This means that the order of lookups for possibly non-existing accounts should always be
1. Read CodeHash
2. If CodeHash != 0, then read other fields as necessary

By impossible what I mean is that it won't be possible to generate a list of MPT updates + proofs that can verify it.  Since we don't have the MPT circuit connected yet, tests will not fail (although we could add some checks to make sure we never read non-codeHash account fields of non-existing accounts).

So I think you should move the callee CodeHash lookup to happen first, and if account exists, then do the callee Balance lookup.  I think this affects the circuits as well.  I see in `error_oog_call.rs` first the callee Balance is read, and then ` call_gadget.gas_cost_expr` is called, which causes the callee CodeHash read.  Actually, I'm not sure why in `error_oog_call` the callee balance is read, there doesn't seem to be any constraint using that value: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/ea6d8e38bcaedbf2d69d0661ad14485c6b02b9cb/zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs#L66-L71

So I think you should change the order of reads in the test for the callee: first read the CodeHash, then if its != 0, read the Balance.

After a second look at the test code you mentioned https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_callop.py#L345-L355 I see that the Balance read in there is for the caller which is independent from the callee CodeHash read; which means that the order doesn't matter, but it's still OK to read the callee CodeHash before the caller Balance.

So in summary: for possibly non-existing accounts, first read codeHash, and only if its !=0 read other fields.  For accounts known to be existing in the context (like the caller in the call context), the order of reads is not important.

Reply:
the order of CodeHash read was fixed in [7e0fc6e](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/348/commits/7e0fc6e7ae1880835beefa735e8fe290338eac15)
.

> Actually, I'm not sure why in error_oog_call the callee balance is read, there doesn't seem to be any constraint using that value:

yep, balance is not necessary in `error_oog_call`, I have removed it.
.

> So I think you should change the order of reads in the test for the callee: first read the CodeHash, then if its != 0, read the Balance.

For `then if its != 0, read the Balance`, I'm not sure I understand it corretly. Do you mean add a condiftion `if callee_bytecode_hash != 0` at [test_callop.py#L349](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_callop.py#L349) and test_callop.py#L353? From my understanding, in evm, user's balance is still able to be read even it's a non-existing account (not initialized in MPT). If we add `if callee_bytecode_hash != 0`, I guess it'll panic(or just fail) if we read the balance of a non-existing account. Not sure my understanding is correct and not sure I understand your meaning correctly.

Reply:
> For `then if its != 0, read the Balance`, I'm not sure I understand it corretly. Do you mean add a condiftion `if callee_bytecode_hash != 0` at [test_callop.py#L349](https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/tests/evm/test_callop.py#L349) and test_callop.py#L353?

I was mistaken on my interpretation of the callop logic, so the `then if its != 0, read the Balance` doesn't apply here, because the Balance is read on the caller, which at this point is guaranteed to exist, so there's no need to make it conditional.  I think there's an edge case on the transfer thought, but more on it later.

> From my understanding, in evm, user's balance is still able to be read even it's a non-existing account (not initialized in MPT).

From the EVM spec, whenever an opcode that reads account fields encounters a non-existing account particular conditions are applied.  Most of the time the behavior is the same as considering the non-existing account as a default account (with nonce=0, balance=0, code_hash=EMPTY_HASH), but that's not always the case.  For example, EXTCODEHASH returns 0 for non-existing accounts (instead of EMPTY_HASH).  This also affects writes, as opcodes that write to accounts may have different gas cost when they create the account (because it was non-existing before).

In our circuits, account read/writes are triggered by the EVM circuit.  Those read/writes are then verified for consistency in the State Circuit, and finally they are collected and matched against MPT proofs to match the state trie transitions.  Our current design of the MPT Circuit supports the proofs listed in this table: https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md#mpt_table

In particular, the `NonceMod`, `BalanceMod` and `CodeHashMod` don't support a read (valuePrev == value) of non-existing accounts.  They only support reads to existing accounts, and writes to existing accounts or to non-existing accounts (which triggers account creation).

So a Nonce, Balance, CodeHash read proof for non-existing accounts is not supported by the MPT circuit.

The MPT circuit supports `NonExistingAccountProof` which acts like a read and proves that an account doesn't exist.

And in the StateCircuit we map `CodeHash` read of 0 to `NonExistingAccountProof` in the MPT.  We can do this because code_hash will never be 0 in an existing account, so wenever in the EVM/State circuit a CodeHash is read as 0, it means that the account doesn't exist.  And if the account doesn't exist, a following Nonce or Balance read (triggered by the EVM) can't be proved by the MPT Circuit.

This is why every time a Nonce/Balance is read from the EVM circuit, we must make sure that the account exists, otherwise it may be translated to a  `NonceMod` or `BalanceMod` proof in the MPT that is impossible to succeed.  So the idea is, in those cases, have a boolean expression that encodes the account existence, and use it to enable/disable the field lookups.

The way we disable a lookup is by:
1- In all tables always have an all 0 row
2- Multiply all expressions of the lookup by the "selector" expression, so that when it's zero, the lookup is a list of 0s and always suceceds, bypassing the other possibly non-zero values in the original lookup expression.

> If we add `if callee_bytecode_hash != 0`, I guess it'll panic(or just fail) if we read the balance of a non-existing account. Not sure my understanding is correct and not sure I understand your meaning correctly.

I'm not sure I understand this part

Reply:
BTW I'm working on a simple check on zkevm-circuits that will guarantee that we don't have invalid account field read/writes: https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1084

Reply:
I consider this conversation resolved!  But we can continue a discussion to resolve any doubts, etc :)